### PR TITLE
Program Interface implementation library for implementing programs

### PR DESCRIPTION
**Pull request by buffalojoec**
After reviewing your latest PR for the Token Metadata Interface, I decided to re-work the stuff I was laying down in [this repo I linked in your PR](https://github.com/buffalojoec/srfc-interface-discriminators) into a draft PR that would actually leverage the instructions from an interface the way you wrote them in Token Metadata @joncinque .

This is only a draft for now, but I'd like to get your thoughts @joncinque @ngundotra

---

The reason I'm sharing now - while it's still a draft - is because I'd like to discuss the formatting for interfaces declared in the SPL repository, and adding only whatever necessary traits/constants/etc. for validation. We can get a little thread going below on ideas.

Mainly to replace the manual implementations here: https://github.com/buffalojoec/solana-program-library-jonc/blob/interface-instructions/libraries/program-interface/syn/src/instructions.rs

---

I encourage you guys to take a look at the README in the linked repo, and also [this test file](https://github.com/buffalojoec/srfc-interface-discriminators/blob/main/interface-instructions/tests/mod.rs) to get an understanding of what I'm shooting for.

The macro-esque code is for Shank & Native, but the underlying functions for validation and discriminators would be leveraged by Anchor.

### CODE REVIEW

**Comment by joncinque** [comment]
This is really super cool, especially when combined with the type-checking, and it works so seamlessly with other tools. It should definitely be used!

We'll need to define behavior for non-framework programs. For example, if you use `BorshDeserialize` on a normal enum, it'll generate 1 byte for the discriminator by default, but then interfaces will get 8 bytes. So what happens if you have something like:

```
enum MyInstruction {
    MyInstruction, // = 0
    #[interface(first_interface)]
    InterfaceIx, // = [8, bytes, for, the, interface, ...]
    MyOtherInstruction, // = 1?
    #[interface(second_interface)],
    SecondInterfaceIx, // = [8 bytes ... ]
}
```

Or keep it simple and enforce that `SplInterfaceInstruction` always gives 8 byte discriminators no matter what. That might be the best solution!

For something like implementing metadata in token-2022, since it uses manual pack / unpack for the existing instructions, it probably wouldn't be able to use this directly, and instead need to specify *another* instruction type for the interfaces.

**Comment by buffalojoec** [comment]
> This is really super cool, especially when combined with the type-checking, and it works so seamlessly with other tools. It should definitely be used!
> 
> We'll need to define behavior for non-framework programs. For example, if you use `BorshDeserialize` on a normal enum, it'll generate 1 byte for the discriminator by default, but then interfaces will get 8 bytes. So what happens if you have something like:
> 
> ```
> enum MyInstruction {
>     MyInstruction, // = 0
>     #[interface(first_interface)]
>     InterfaceIx, // = [8, bytes, for, the, interface, ...]
>     MyOtherInstruction, // = 1?
>     #[interface(second_interface)],
>     SecondInterfaceIx, // = [8 bytes ... ]
> }
> ```
> 
> Or keep it simple and enforce that `SplInterfaceInstruction` always gives 8 byte discriminators no matter what. That might be the best solution!
> 
> For something like implementing metadata in token-2022, since it uses manual pack / unpack for the existing instructions, it probably wouldn't be able to use this directly, and instead need to specify _another_ instruction type for the interfaces.

Yeah, I figured we could just follow Anchor's pattern by implementing the discriminator for any user-defined non-interface instructions as `hash("global:" + <ix name>)[..8]`. This way, all instructions have 8 bytes. 

As far as Borsh goes, I figured we'd want to just do a pack/unpack like you're doing with Token2022, and then devs would just have to just use `let ix = MyInstructions::unpack(&data)?` and then match on `ix`

I guess my goal here was twofold:
1. Discuss the drawbacks of enforcing use of an enum for instruction layout, with subsequent processor matching (which seems to be pretty de facto)
2. Ideate on how to best dynamically represent SPL defined interfaces as `syn::Type`-based validation structs, for validating with implementing program code (this is difficult to do without a macro inside a macro, but I guess that option isn't off the table, eh?)

**Comment by ngundotra** [comment]
Can you highlight how this would be used? That will help me go backwards & provide feedback

**Comment by buffalojoec** [comment]
> Can you highlight how this would be used? That will help me go backwards & provide feedback

Check out the program specs:
- [Anchor](https://github.com/buffalojoec/srfc-interface-discriminators/blob/main/programs/anchor-example/src/lib.rs)
- [Native](https://github.com/buffalojoec/srfc-interface-discriminators/blob/main/programs/native-example/src/lib.rs)
- [Shank](https://github.com/buffalojoec/srfc-interface-discriminators/blob/main/programs/shank-example/src/lib.rs) (which is the same as Native)

The idea is that in Anchor, all you need is the annotations over your functions, and in Shank/Native it's the same plus use of the `MyInstruction::unpack(..)` in the processor, to implement one or more interfaces!

It's even got some nice little errors for implementing, so there's virtually no glass-chewing: https://github.com/buffalojoec/srfc-interface-discriminators/blob/main/README.md

---

TL/DR: This macro lib will validate you are implementing the interface exactly as you are supposed to, and preface your instructions with the various interface or `global:<>` discriminators.
